src/client/datascience/kernel-launcher/jupyterPaths.ts

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+
+import { inject, injectable } from 'inversify';
+import * as path from 'path';
+import { CancellationToken } from 'vscode';
+import { IPlatformService } from '../../common/platform/types';
+import { IPathUtils } from '../../common/types';
+import { IEnvironmentVariablesProvider } from '../../common/variables/types';
+import { tryGetRealPath } from '../common';
+
+const winJupyterPath = path.join('AppData', 'Roaming', 'jupyter', 'kernels');
+const linuxJupyterPath = path.join('.local', 'share', 'jupyter', 'kernels');
+const macJupyterPath = path.join('Library', 'Jupyter', 'kernels');
+export const baseKernelPath = path.join('share', 'jupyter', 'kernels');
+
+@injectable()
+export class JupyterPaths {
+    constructor(
+        @inject(IPlatformService) private platformService: IPlatformService,
+        @inject(IPathUtils) private readonly pathUtils: IPathUtils,
+        @inject(IEnvironmentVariablesProvider) private readonly envVarsProvider: IEnvironmentVariablesProvider
+    ) {}
+
+    /**
+     * This should return a WRITABLE place that jupyter will look for a kernel as documented
+     * here: https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
+     */
+    public async getKernelSpecRootPath(): Promise<string | undefined> {
+        if (this.platformService.isWindows) {
+            return tryGetRealPath(path.join(this.pathUtils.home, winJupyterPath));
+        } else if (this.platformService.isMac) {
+            return path.join(this.pathUtils.home, macJupyterPath);
+        } else {
+            return path.join(this.pathUtils.home, linuxJupyterPath);
+        }
+    }
+    /**
+     * This list comes from the docs here:
+     * https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs
+     */
+    public async getKernelSpecRootPaths(cancelToken?: CancellationToken): Promise<string[]> {
+        // Paths specified in JUPYTER_PATH are supposed to come first in searching
+        const paths: string[] = await this.getJupyterPathPaths(cancelToken);
+
+        if (this.platformService.isWindows) {
+            const winPath = await this.getKernelSpecRootPath();
+            if (winPath) {
+                paths.push(winPath);
+            }
+
+            if (process.env.ALLUSERSPROFILE) {
+                paths.push(path.join(process.env.ALLUSERSPROFILE, 'jupyter', 'kernels'));
```

Comment:
I think you might need tryGetRealPath on all paths returned on windows.

Reply:
Lets do that separately

Reply:
I'll file an issue for ath.

Reply:
Not sure we need to do that at all, I thought the environment variable contains the fully qualified path.

Reply:
Added an issue here https://github.com/microsoft/vscode-jupyter/issues/6352